### PR TITLE
feat: new crate for no-std groth/plonk verification

### PR DESCRIPTION
**Pull request by yuwen01**
New crate for lightweight verification of groth16/plonk proofs generated by sp1. Usable in sp1 zkvm and wasm contexts.

Copied + adapted from [this repo](https://github.com/succinctlabs/snark-bn254-verifier/tree/main)

Changes from original repo:

1. Feature flagged plonk behind the getrandom flag, so we can still verify groth16 proofs in fully no-std environments.
2. Used an upgraded version of the [bn patch](https://github.com/sp1-patches/bn) that feature flags using the `sp1-lib` to activate precompiles. We need other functionality from this patch besides the zkvm precompile activation stuff, and we can only use the zkvm with the standard library. 
3. Miscellaneous cleaning

Also
1. Added a groth16 example, which generates a fibonacci proof and verifies it in sp1. 
2. Added another method in the SDK which is needed to generate proof strings for plonk proofs. `.bytes()` abi encodes the proof, which is only useful in ethereum contexts.

### CODE REVIEW

File: crates/sdk/src/proof.rs [hunk 1]
```diff
@@ -53,6 +53,32 @@ impl SP1ProofWithPublicValues {
...
+            SP1Proof::Plonk(plonk) => {
+                let mut bytes = Vec::with_capacity(4 + (plonk.raw_proof.len()) / 2);
+                bytes.extend_from_slice(&plonk.plonk_vkey_hash[..4]);
+                bytes.extend_from_slice(
+                    &hex::decode(&plonk.raw_proof).expect("Invalid Plonk proof"),
+                );
+                bytes
+            }
+            SP1Proof::Groth16(groth16) => {
+                let mut bytes = Vec::with_capacity(4 + (groth16.raw_proof.len()) / 2);
```
[the remark this answers is not in the document]
Comment (yuwen01) [reply crates/sdk/src/proof.rs:73]: added more comments so hopefully it's more clear. We're turning a hex string into bytes, so 2 hex characters is one byte. 

This gets the raw proof, not the abi-encoded proof like `bytes()`. It doesn't matter for groth16, but it does for plonk.

[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/sdk/src/proof.rs:73]: Why can't you just do:

```rust
let proof_bytes = hex::decode(plonk.raw_proof)
concat!(plonk.plonk_vkey_hash[..4], proof_bytes);
```

[the remark this answers is not in the document]
Comment (ratankaliani) [reply crates/sdk/src/proof.rs:73]: That's more straightforward and easier to parse

File: crates/verifier/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# SP1 Verifier
+
+This crate provides verifiers for SP1 Groth16 and Plonk zero-knowledge proofs. These proofs are expected
+to be generated using the [SP1 SDK](../sdk).
+
+## Features
+
+Groth16 and Plonk proof verification are supported in completely no-std environments.
```
Comment (ratankaliani) [inline crates/verifier/README.md:8]: rm "completely"

Make no-std, `no-std`.

File: crates/verifier/src/constants.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/// Gnark (and arkworks) use the 2 most significant bits to encode the flag for a compressed point.
```
Comment (ratankaliani) [inline crates/verifier/src/constants.rs:1]: Docs: Add a reference for this?

File: crates/verifier/src/converter.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    } else {
+        let mut x_bytes: [u8; 32] = [0u8; 32];
+        x_bytes.copy_from_slice(buf);
+        x_bytes[0] &= !MASK;
+
+        let x = Fq::from_be_bytes_mod_order(&x_bytes).expect("Failed to convert x bytes to Fq");
+
+        Ok((x, m_data.into()))
+    }
+}
```
Comment (ratankaliani) [inline crates/verifier/src/converter.rs:48]: The interface for `is_zeroed` can be simplified to do the mask inside of the loop, and the logic for the loop can also be simplified with:

```rust
    first_byte == 0 && buf.iter().all(|&b| b == 0)
```

File: crates/verifier/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    #[error("Challenge not found")]
+    ChallengeNotFound,
+    #[error("Previous challenge not computed")]
+    PreviousChallengeNotComputed,
+    #[error("Pairing check failed")]
+    PairingCheckFailed,
+    #[error("Invalid point in subgroup check")]
+    InvalidPoint,
+
+    // Mathematical Errors
```
Comment (ratankaliani) [inline crates/verifier/src/error.rs:20]: docs: Arithmetic Errors

File: crates/verifier/src/groth16/converter.rs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    let ar = uncompressed_bytes_to_g1_point(&buffer[..64])?;
+    let bs = uncompressed_bytes_to_g2_point(&buffer[64..192])?;
+    let krs = uncompressed_bytes_to_g1_point(&buffer[192..256])?;
+
+    Ok(Groth16Proof { ar, bs, krs })
+}
+
+pub(crate) fn load_groth16_verifying_key_from_bytes(
+    buffer: &[u8],
+) -> Result<Groth16VerifyingKey, Groth16Error> {
```
Comment (ratankaliani) [inline crates/verifier/src/groth16/converter.rs:23]: docs: Add description for both functions

File: crates/verifier/src/groth16/verify.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    Ok(public_inputs
+        .iter()
+        .zip(vk.g1.k.iter().skip(1))
+        .fold(vk.g1.k[0], |acc, (i, b)| acc + (*b * *i))
+        .into())
+}
+
+pub(crate) fn verify_groth16_raw(
```
Comment (ratankaliani) [inline crates/verifier/src/groth16/verify.rs:45]: docs:

File: crates/verifier/src/groth16/verify.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+pub(crate) struct Groth16Proof {
+    pub(crate) ar: AffineG1,
+    pub(crate) krs: AffineG1,
+    pub(crate) bs: AffineG2,
+}
+
+// Prepare the inputs for the Groth16 verification by combining the public inputs with the
+// corresponding elements of the verification key.
+fn prepare_inputs(vk: Groth16VerifyingKey, public_inputs: &[Fr]) -> Result<G1, Groth16Error> {
```
Comment (ratankaliani) [inline crates/verifier/src/groth16/verify.rs:33]: Make these a Rust doc comment

File: crates/verifier/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,34 @@
+//! This crate provides verifiers for SP1 Groth16 and Plonk BN254 proofs in a no-std environment.
+//! It is patched for efficient verification within the SP1 ZKVM context.
+#![deny(rustdoc::broken_intra_doc_links)]
+#![deny(missing_debug_implementations)]
+#![deny(missing_docs)]
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (ratankaliani) [inline crates/verifier/src/lib.rs:6]: Should these deny! implementations still be present? IMO you should remove them and then add RustDoc's for all of the functions/structs.

File: crates/verifier/src/plonk/converter.rs
Comment (ratankaliani) [inline crates/verifier/src/plonk/converter.rs:1]: missing function docs

File: crates/verifier/src/plonk/hash_to_field.rs
Comment (ratankaliani) [inline crates/verifier/src/plonk/hash_to_field.rs:1]: missing some function docs

File: crates/verifier/bn254-vk/groth16_vk.bin
Comment (ratankaliani) [inline crates/verifier/bn254-vk/groth16_vk.bin:1]: How is the `groth16_vk.bin` generated? Is it from each SP1 release? Do you have an automatic script which fetches this?

Comment (ratankaliani) [reply crates/verifier/bn254-vk/groth16_vk.bin:1]: Same for `plonk_vk.bin`.

File: crates/verifier/src/plonk/verify.rs
Comment (ratankaliani) [inline crates/verifier/src/plonk/verify.rs:1]: function docs

**Review by ratankaliani — COMMENTED** [review]
Left some comments for code simplification and missing docs. You should also add instructions in the README for how the `groth16_vk.bin` and `plonk_vk`.bin are generated.

File: crates/verifier/bn254-vk/groth16_vk.bin
Comment (yuwen01) [reply crates/verifier/bn254-vk/groth16_vk.bin:1]: For the record, they're already being checked in ci (as a unit test in tests.rs). I'll update some of the release scripts to copy the vkeys to this directory too.

File: crates/verifier/src/plonk/verify.rs [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+        Some(zu.into_u256().to_bytes_be().to_vec()),
+        &mut fs,
+    )?;
+
+    // Derives the final randomness U.
+    let u = derive_randomness(
+        &mut fs,
+        U,
+        Some(vec![folded_digest, proof.z, folded_proof.h, proof.z_shifted_opening.h]),
+    )?;
```
Comment (yuwen01) [inline crates/verifier/src/plonk/verify.rs:301]: @rkm0959 updated fs with your fixes <3

**Pull request merged**